tests/executables_src/testDataValidityPropagation.cc

```diff
@@ -0,0 +1,552 @@
+  // redmine issue: #8550
+}
+
+/*
+ * \anchor testDataValidity_1_4 \ref dataValidity_1_4 "1.4"
+ * The user code has the possibility to query the data validity of the module
+ *
+ * No explicit test, if test suite compiles it's ok.
+ */
+
+/*
+ * \anchor testDataValidity_1_5 \ref dataValidity_1_5 "1.5"
+ * The user code has the possibility to set the data validity of the module to 'faulty'. However, the user code cannot
+ * actively set the module to 'ok' if any of the module inputs are 'faulty'.
+ *
+ * No explicit test. The module should use increment/decrement mechanism to set invalid state,
+ * which implies it cannot override faulty state.
+ * BUT it is actually possible to override getDataValidity in the Module, so spec does not hold strictly!
+ *
+ */
+
+struct TestApplication16 : ctk::Application {
+  TestApplication16() : Application("testSuite") {}
+  ~TestApplication16() { shutdown(); }
+
+  void defineConnections() {
+    mod1("o1") >> mod2("i1");
+    findTag(".*").connectTo(cs);
+  }
+
+  TestModule1 mod1{this, "m1", ""};
+  TestModule1 mod2{this, "m2", ""};
+  ctk::ControlSystemModule cs;
+};
+
+/**
+ * \anchor testDataValidity_1_6 \ref dataValidity_1_6 "1.6"
+ * The user code can flag individual outputs as bad. However, the user code cannot actively set an output to 'ok' if
+ * the data validity of the module is 'faulty'.
+ */
+BOOST_AUTO_TEST_CASE(testDataValidity_1_6) {
+  TestApplication16 app;
+  app.mod1.outputValidity = ctk::DataValidity::faulty;
+  app.mod2.outputValidity = ctk::DataValidity::ok;
+  ctk::TestFacility test;
+
+  //app.dumpConnections();
+  auto input = test.getScalar<int>("/m1/i1");
+  auto result = test.getScalar<int>("/m2/o1");
+  test.runApplication();
+  input.write();
+  test.stepApplication();
+  input.write();
+  test.stepApplication();
+
+  // module 2 must be flagged bad because of the faulty input from module 1
+  BOOST_CHECK(app.mod2.getDataValidity() == ctk::DataValidity::faulty);
+  result.read();
+  // output of module 2 cannot be valid, even if module tries to set it to valid
+  BOOST_CHECK(result.dataValidity() == ctk::DataValidity::faulty);
+}
+
+/*
+ * \anchor testDataValidity_1_7 \ref dataValidity_1_7 "1.7"
+ *  The user code can get the data validity flag of individual inputs and take special actions.
+ *
+ *  No explicit test required.
+ */
+
+/**
+ * \anchor testDataValidity_1_8 \ref dataValidity_1_8 "1.8"
+ * The data validity of receiving variables is set to 'faulty' on construction. Like this, data is marked as faulty as
+ * long as no sensible initial values have been propagated.
+ */
+BOOST_AUTO_TEST_CASE(testDataValidity_1_8) {
+  TestApplication1<TestModule0> app;
+  // testable mode cannot be used here, since it would wait on initial values (which are not provided)
+  ctk::TestFacility test(false);
+
+  auto o0 = test.getScalar<int>("/m1/oNothing");
+  test.runApplication();
+
+  // o0.read() would block here
+  o0.readLatest();
```

Comment:
I would write:
```
BOOST_CHECK(o0.readNonBlocking() == false);
```
(and keep the other `BOOST_CHECK` below of course)

Reply:
ok